File: java/src/main/java/com/mlt/vector/constant/ConstVector.java
package com.mlt.vector.constant;

import com.mlt.vector.BitVector;
import com.mlt.vector.Vector;
import java.nio.Buffer;

public abstract class ConstVector<T extends Buffer, K> extends Vector<T, K> {

  public ConstVector(String name, T buffer, int size) {
    super(name, buffer, size);
    this.size = size;
  }

  public ConstVector(String name, BitVector nullabilityBuffer, T buffer) {
    super(name, nullabilityBuffer, buffer);
  }
}

File: ts/src/vector/constant/constVector.ts
import BitVector from "../flat/bitVector";
import Vector from "../vector";
import {SelectionVector} from "../filter/selectionVector";
import {FlatSelectionVector} from "../filter/flatSelectionVector";


export abstract class ConstVector<T extends ArrayBuffer, K> extends Vector<T, K> {

    protected constructor(name: string, buffer: T, sizeOrNullabilityBuffer: number | BitVector){
        super(name, buffer, sizeOrNullabilityBuffer);
    }

    protected updateSelectionVector(selectionVector: SelectionVector) {
        const buffer = selectionVector.selectionValues();
        let limit = 0;
        for (let i = 0; i < selectionVector.limit; i++) {
            if (!this.nullabilityBuffer || this.nullabilityBuffer.get(i)) {
                buffer[limit++] = buffer[i];
            }
        }

        selectionVector.setLimit(limit);
    }

    protected createSelectionVector() {
        const selectionVector = [];
        for (let i = 0; i < this.size; i++) {
            if (!this.nullabilityBuffer || this.nullabilityBuffer.get(i)) {
                selectionVector.push(i);
            }
        }

        return new FlatSelectionVector(selectionVector);
    }
}
